### ISSUE TITLE
[Love 12.0] Buffer with floatmat4x4 in format returns an error with setArrayData()

### ISSUE THREAD

**Issue by Froyok** [love2d/love#1986]
Initially I wanted to build a Buffer with a custom format, something like:
```lua
local Format= {
    { name="Time", format="float" },
    { name="CameraProjMatrix", format="floatmat4x4" },
    { name="FogColor", format="floatvec3" }
}
```
But Love 12 returns an error when calling `setArrayData`, something along those lines:
`Array length in flat array variant of Buffer:setArrayData must be a multiple of the total number of components (31)`

After discussing this on Discord with Sasha (@slime73), I also tried the following commit which quickly added the support for mat4: https://github.com/love2d/love/commit/910533eddaadd7bef3f86da717f4a8a9dee18704 (tested via this action: https://github.com/love2d/love/actions/runs/6983798221).

However using this new build lead to another error: `calling 'setArrayData' on bad self (number expected, got userdata)`

Looking into the issue a bit further, it is related to mat4, switching to a mat2 works (mat3 aren't supported still). This issue seems to only appear on my Linux configuration, on Windows everything looks fine at first glance (no error but I haven't checked the Buffer values).

This behavior was observed in a complex project, with the following script there is still an error but a different one:
```lua
local Buffer = nil

function love.load( Args )
    local Format = {
        { name="1", format="float" },
        { name="2", format="floatmat4x4" },
        { name="3", format="floatvec2" }
    }
    -- 1 + 16 + 2 = 19

    local Temp = {}

    for i=1, 19 do
        Temp[i] = 0
    end

    local Buffer = love.graphics.newBuffer(
        Format,
        1,
        { shaderstorage = true }
    )

    print( Buffer:getElementStride() )

    Buffer:setArrayData( Temp )
end

function love.draw()
end
```

This script produces the following output:
```bash
$ ./run.sh 
96
Error: main.lua:25: bad argument #-1 to 'setArrayData' (number expected, got number)
stack traceback:
        [love "boot.lua"]:431: in function <[love "boot.lua"]:427>
        [C]: in function 'setArrayData'
        main.lua:25: in function 'load'
        [love "callbacks.lua"]:150: in function <[love "callbacks.lua"]:149>
        [C]: in function 'xpcall'
        [love "boot.lua"]:447: in function <[love "boot.lua"]:434>
        [C]: in function 'xpcall'
```

System and Love version:

- Love 12
- LuaJIT 2.1.1696795921
- OS: Linux Manjaro
- Arch: x64
- Renderer Name: OpenGL
- Renderer Version: 4.6 (Core Profile) Mesa 23.1.5
- Renderer Vendor: AMD
- Renderer Device: AMD Radeon RX 7900 XT (gfx1100, LLVM 15.0.7, DRM 3.52, 6.4.9-1-MANJARO)

As mentionned, this was discussed on discord here:

- Initial message: https://discord.com/channels/329400828920070144/329404715521802241/1177666599239549038
- Test of new commit: https://discord.com/channels/329400828920070144/329404715521802241/1177691710302732289
- Debug thread: https://discord.com/channels/329400828920070144/1177731018548064327/1177731020766846987

**Comment by slime73** [comment]
I added a test to our new testing framework for this. The good news is that it [reproduces](https://github.com/love2d/love/actions/runs/7372484175) on Github Actions' Linux machines (but not on Windows/macOS). The bad news is I still don't know why it happens. I'll investigate more...

**Comment by slime73** [comment]
Nevermind, I found the issue - there was some Buffer component wrapper code that accessed a static 4 element array so it read out-of-bounds memory when reading a 16 element 4x4 matrix. It's fixed now in [9b537ed](https://github.com/love2d/love/commit/9b537ede2b18dadf30371412eeeddda0dd57fb8a).

**Comment by Froyok** [comment]
I confirm it seems to be working as intended, I replaced my old workaround code with the expected behavior and everything looks normal. :)
